### ISSUE TITLE
reset by path

### ISSUE THREAD

**Issue by jennybc** [ropensci/git2r#251]
Related to, but unfortunately not solved by #250 which exposed `git_index_remove_bypath` (thanks!).

Reminder of context: With command line Git, it's possible to undo `git add` for specific paths. To quote the [`git reset` documentation](https://git-scm.com/docs/git-reset):

> `git reset <paths>` is the opposite of `git add <paths>`

It turns out that simply removing a path from the index with the new `index_remove_bypath()` is only the right thing to do when the path was previously untracked. I found a description of how to unstage in general in (https://github.com/libgit2/libgit2/issues/3632):

> The act of "unstaging" is to update the repository's index to match the HEAD tree. So you don't want to remove anything from the index, you want to add an index entry that matches HEAD.
>
> That means that you'll need to reset to something. You should lookup the HEAD commit and reset to that.
>
> For example:
>
> ```
> git_reference *head;
> git_object *head_commit;
>
> git_repository_head(&head, repo));
> git_reference_peel(&head_commit, head, GIT_OBJ_COMMIT);
> git_reset_default(repo, head_commit, &paths);
> ```

For the record, I did not have as negative an experience as the OP on the `libgit2` thread. From a single test, it seems like using the new `index_remove_bypath()` on a previously tracked file leaves the file contents as they should be. But with an odd Git status. It's reported to be "unmerged", when it should be "unmodified".

``` r
library(git2r)
repo <- tempfile("git-unstage-example-")
dir.create(repo)
gr <- init(repo)
owd <- setwd(repo)
write("line 1", "unstage-me.txt")
status(gr)
#> Untracked files:
#>  Untracked:  unstage-me.txt
add(gr, "unstage-me.txt")
commit(gr, message = "first commit")
#> [5235df2] 2016-08-07: first commit

write("and another line", "unstage-me.txt", append = TRUE)
## this is the status I expect to see after unstaging
status(gr)
#> Unstaged changes:
#>  Modified:   unstage-me.txt

add(gr, "unstage-me.txt")
status(gr)
#> Staged changes:
#>  Modified:   unstage-me.txt
readLines("unstage-me.txt")
#> [1] "line 1"           "and another line"
index_remove_bypath(gr, "unstage-me.txt")
readLines("unstage-me.txt")
#> [1] "line 1"           "and another line"
status(gr)
#> Unstaged changes:
#>  Unmerged:   unstage-me.txt
setwd(owd)
```

Q1: What do you think of implementing path-specific unstage according to the sketch from the libgit2 issue thread?

Q2: Assuming `git2r` isn't going to change quickly (or at all), what do you think is the best current way to unstage specific paths? Right now I'm doing a mixed reset, then re-staging everything but the requested paths. I wonder if there's a better way, possibly inspired by the libgit2 thread....

**Comment by stewid** [comment]
I have implemented the path-specific unstage according to the sketch from the libgit2 issue thread. I added a new `reset` method with a `path` argument. This method updates the index entries for all paths to their state at HEAD. I think we can drop the `index_remove_bypath` method since the new `reset` method checks which libgit2 function to use depending on if the repository is empty https://github.com/ropensci/git2r/blob/master/R/reset.r#L110
